helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -70,15 +70,29 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
 
-    // Compute overall utilization of the cluster. Capacity dimension -> total remaining capacity
-    Map<String, Integer> overallClusterRemainingCapacityMap =
-        computeOverallClusterRemainingCapacity(nodes);
+    // create a always >0 capacity map to avoid divide by 0.
+    Map<String, Float> positiveEstimateClusterRemainCap = new HashMap<>();
+    for (Map.Entry<String, Integer> clusterRemainingCap : clusterModel.getContext()
+        .getEstimateUtilizationMap().entrySet()) {
+      String capacityKey = clusterRemainingCap.getKey();
+      if (clusterRemainingCap.getValue() < 0) {
+        // all replicas' assignment will fail if there is one dimension's remain capacity <0.
+        throw new HelixRebalanceException(String
+            .format("The cluster does not have enough %s capacity for all partitions. ",
+                capacityKey), HelixRebalanceException.Type.FAILED_TO_CALCULATE);
+      }
+      // estimate remain capacity after assignment + %1 of current cluster remain capacity before assignment
+      positiveEstimateClusterRemainCap.put(capacityKey,
+           clusterModel.getContext().getEstimateUtilizationMap().get(capacityKey) +
+          (clusterModel.getContext().getClusterRemainCapacityMap().get(capacityKey) * 0.01f));
```

Comment:
One more thing, let's have a constant for this magic number

Reply:
TFTR. Updated.